src/posts/2025-09-16-kernel-basics-for-tetragon/index.md

```diff
@@ -0,0 +1,224 @@
+
+### Tetragon Points of View
+
+Tetragon abstracts these raw attachment points into policy targets. **A hook point in Tetragon is simply a declaration of where you want to monitor and what arguments you want extracted.**
+With a kprobe spec for example, you declare, and Tetragon takes care of attaching to the kernel function and extracting the arguments.
+
+```yaml
+spec:
+  kprobes:
+    - call: 'fd_install'
+      syscall: false
+```
+
+With a tracepoint spec, Tetragon subscribes to the stable kernel tracepoints, decoding the arguments.
+
+```yaml
+spec:
+  tracepoints:
+    - subsystem: 'sched'
+      event: 'sched_process_exec'
+```
+
+This same principle applies to every hook point in Tetragon. With LSM hooks, for example, you point at security checks(`file_open`, `bprm_check_security`) and Tetragon handles the attachment.
+
+From the Tetragon point of view, **a hook point is a declarative contract that specifies what to watch(functions, tracepoints, LSM hooks), which arguments to pull out(pid, file, etc), and how to filter or act on the event.**
+
+### Writing Kprobe-based Tracing Policies for Monitoring System Calls
+
+Before rounding off this section, it is important to highlight one interesting abstraction Tetragon provides for kprobe-based policies that monitor system calls. Different CPU architectures implement system calls differently, and this can create portability challenges. On x86_64, system call handlers have names like `__x64_sys_write`, while on ARM64, they're named `__arm64_sys_write`.
+Tetragon provides elegant abstraction here. Instead of writing architecture-specific policies like this:
+
+```yaml
+# Architecture-specific (don't do this)
+spec:
+  kprobes:
+    - call: '__x64_sys_write' # Only works on x86_64
+      syscall: true
+```
+
+You can write portable policies that work across different architectures.
+
+```yaml
+# Portable across architectures
+spec:
+  kprobes:
+    - call: 'sys_write' # Works on any architecture
+      syscall: true
+```
+
+Tetragon automatically translates `sys_write` to the correct architecture-specific function name. This abstraction is crucial for policies that need to work across diverse environments.
+
+### Choosing the Right Hook Point
+
+The best hook point for your policy depends on your specific security observability objective and how much stability you need across environments. If your goal prioritizes portability, tracepoints are often the safest choice. They are built into the kernel source and tend to remain stable across kernel versions.
+If you’re writing security-sensitive policies, LSM hooks are a better option. Because they operate on kernel-owned memory after user input has already been validated, they naturally avoid time-of-check to time-of-use (TOCTOU) pitfalls.
```

Comment:
```suggestion
If you’re writing security-sensitive policies, LSM hooks are a better option. Because they operate on kernel-owned memory after user input has already been validated, they naturally avoid time-of-check to time-of-use (TOCTOU) pitfalls.

```